projects/driver_controls/src/fsm/pedal_calibration_fsm.c

```diff
@@ -0,0 +1,98 @@
+
+  // User can go to validation stage if already calibrated.
+  FSM_ADD_TRANSITION(INPUT_EVENT_PEDAL_CALIBRATION_VALIDATE, state_validate);
+}
+
+FSM_STATE_TRANSITION(state_full_brake) {
+  // The only available next step is full throttle state.
+  FSM_ADD_TRANSITION(INPUT_EVENT_PEDAL_CALIBRATION_FULL_THROTTLE, state_full_throttle);
+
+  // Or repeating full brake stage if pedal was not in the correct position at some point.
+  FSM_ADD_TRANSITION(INPUT_EVENT_PEDAL_CALIBRATION_FULL_BRAKE, state_full_brake);
+}
+
+FSM_STATE_TRANSITION(state_full_throttle) {
+  // The only available next step is validate state. The user must validate the calibration.
+  FSM_ADD_TRANSITION(INPUT_EVENT_PEDAL_CALIBRATION_VALIDATE, state_validate);
+
+  // Or repeating full throttle stage if pedal was not in the correct position at some point.
+  FSM_ADD_TRANSITION(INPUT_EVENT_PEDAL_CALIBRATION_FULL_THROTTLE, state_full_throttle);
+}
+
+FSM_STATE_TRANSITION(state_validate) {
+  FSM_ADD_TRANSITION(INPUT_EVENT_PEDAL_CALIBRATION_VALIDATE, state_validate);
+  FSM_ADD_TRANSITION(INPUT_EVENT_PEDAL_CALIBRATION_START, state_start);
+}
+
+// Output function for the start state.
+static void prv_start_output(FSM *fsm, const Event *e, void *context) {
+  LOG_DEBUG("You are at the menu of pedal calibration.\n");
+}
+
+// Output function for the full brake state.
+// Reads and collects pedal inputs assuming pedal is in full brake position.
+static void prv_full_brake_output(FSM *fsm, const Event *e, void *context) {
+  LOG_DEBUG("Keep the pedal at full brake position.\n");
+  PedalCalibrationStorage *storage = context;
+  pedal_calibration_process_state(storage, PEDAL_CALIBRATION_STATE_FULL_BRAKE);
+  LOG_DEBUG("Full brake stage complete.\n");
+}
+
+// Output function for the full throttle state.
+// Reads and collects pedal inputs assuming pedal is in full throttle position.
+static void prv_full_throttle_output(FSM *fsm, const Event *e, void *context) {
+  LOG_DEBUG("Keep the pedal at full throttle position.\n");
+  PedalCalibrationStorage *storage = context;
+  pedal_calibration_process_state(storage, PEDAL_CALIBRATION_STATE_FULL_THROTTLE);
+  LOG_DEBUG("Full throttle stage complete.\n");
+}
+
+// Output function for the validate state.
+// Note that the calculations on the collected data takes place here.
+// Validation process continues until fsm has a new transition.
+// It constantly logs the current zone info of the pedal.
+static void prv_validate_output(FSM *fsm, const Event *e, void *context) {
+  PedalCalibrationStorage *storage = context;
+  pedal_calibration_calculate(storage, storage->throttle_calibration_data);
+
+  LOG_DEBUG("Use the pedal to validate the calibration.\n");
+
+  ThrottleStorage throttle;
+  ThrottlePosition position = { .zone = THROTTLE_ZONE_BRAKE, .numerator = 0 };
+  throttle_init(&throttle, storage->throttle_calibration_data, storage->ads1015_storage);
+
+  Event new_event;
+  do {
+    event_process(&new_event);
```

Comment:
`event_process` doesn't always return a valid event. Why do you have an event processing loop in the output of an FSM entry function?